usb: Handle custom request not only for interface zero

eobalski opened the pull request:
This commit reduce the limitation in the USB core stack.
Until now custom handler could only be served if the request
was directed to the first interface associated with the USB device.

This limitation allows USB audio class to properly respond to the
requests.

This is first step to properly handle set/get interface request by the USB audio class.
The exception provided shall not be considered valid solution for all classes.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

carlescufi commented on subsys/usb/usb_device.c:
```diff
@@ -1226,9 +1226,7 @@ static int custom_handler(struct usb_setup_packet *pSetup,
 		}
 
-		if ((iface->custom_handler) &&
-		    (if_descr->bInterfaceNumber ==
```
Please check that all custom handlers behave correctly if the interface number is not intended for it.

emob-nordic replied:
I've checked current status of class implementation in terms of using *custom_handler*.
3 classes require attention and I would like to speak about it sync meeting.

- [ ] [hid class](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/usb/class/hid/core.c#L491) There is no problem with limitation reduction. Class custom handler checks if the requested interface number matches registered hid devices.

- [ ] [dfu class](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/usb/class/usb_dfu.c#L657) Similar solution as hid has could be implemented but the question is if this is relevant? Shouldn't dfu be single-function USB device? I'd like to discuss.
- [ ] [web usb](https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/subsys/usb/webusb/src/webusb.c#L86) Same concerns as above + this is implemented as sample, not as class.

carlescufi commented on include/usb/usb_device.h:
```diff
@@ -112,6 +112,9 @@ typedef void (*usb_ep_callback)(u8_t ep,
  * set by the callback function with the length and the address of the
  * data to be transmitted buffer respectively.
+ *
+ * @return 0 on success, positive value if request is intended to be handled
```
@return 0 on success, `-EINVAL` if the request has not been handled by the custom handler and instead needs to be handled by the core USB stack. Any other error code to denote failure within the custom handler.

emob-nordic replied:
I took this from callback doc and put only for `custom_handler` when we need that. No sense to change every type of handlers there.

Commit:
usb: Allow Audio Class to handle custom_hander for all interfaces

This exception allows USB Audio Class to properly respond to the
interface requests.

This commit is temporary solution and shall not be considered as
valid solution for other classes.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

Commit:
usb: Special return values for custom_handler

This commit introduces dedicated return type for custom_handler.
Relevant code is updated to fulfill the API documentation.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

emob-nordic commented:
@anangl, @jfischer-phytec-iot, @carlescufi  Please revisit.

emob-nordic commented:
Last push - fixed checkpatch issues.

carlescufi commented on subsys/usb/usb_device.c:
```diff
@@ -1226,9 +1226,13 @@ static int custom_handler(struct usb_setup_packet *pSetup,
+		/* An exception for AUDIO_CLASS is temporary and shall not be
+		 * considered as valid solution for other classes.
+		 */
+		if (iface->custom_handler &&
```
@emob-nordic @jfischer-phytec-iot 
Mind you why not have 2 handlers here instead of the hack with `== AUDIO`?
You could have:

```
if (iface->custom_handler_all) {
   return iface->custom_handler_all(pSetup, len, data);
}
else if ((iface->custom_handler) && (if_descr->bInterfaceNumber == sys_le16_to_cpu(pSetup->wIndex)))
{
   return iface->custom_handler(pSetup, len, data);
}
```

jfischer-phytec-iot replied:
Well add an other handler is API addition and there are a few reasons to rework how the stack interfaces with classes and to remove something again from the API is a more complicated process as just to add something. I suggest to go with the hack but revise the interfacing after release.

emob-nordic replied:
I agree with Johann. For this release let's follow this hack. I need to rethink what could be done to improve the interface between core stack and classes.

carlescufi reviewed (approved):
Accepted as it is.
